test-data/unit/check-overloading.test

```diff
@@ -2188,33 +2188,42 @@ def bar2(*x: int) -> int: ...
 from typing import overload, TypeVar, Generic
 
 T = TypeVar('T')
+# The examples below are unsafe, but it is a quite common pattern
+# so we ignore the possibility of type variables taking value `None`
+# for the purpose of overload overlap checks.
 
 @overload
-def foo(x: None, y: None) -> str: ...  # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
+def foo(x: None, y: None) -> str: ...
 @overload
 def foo(x: T, y: T) -> int: ...
 def foo(x): ...
```

Comment:
Test also calling this to make sure we infer the return type correctly?

Reply:
Oh well, this discovered an obvious flaw in my solution, for this
```python
@overload
def foo(x: None) -> None: ...
@overload
def foo(x: T) -> list[T]: ...

x: int | None
foo(x)
```
we will continue to infer `list[int | None]`, not `list[int] | None` that people will likely expect. I will try to check if there is a simple way to force union-math for this case. If there are none, this PR may become much more complex.

Reply:
OK, I think I have come up with some heuristics to force union math for `None` vs `TypeVar` overlap situations. It is not very precise, but should be generally safe, since IIUC union math infer better types anyway, we usually try to avoid it simply because it is computationally intensive.